#include "mainwindow.h"
#include "ui_mainwindow.h"
#include <QStyleFactory>
#include <QRandomGenerator>
#include <QMessageBox>
#include <QTranslator>
#include <QFile>
#include <QCoreApplication>
#include <QSqlDatabase>
#include <QSqlQuery>
#include <QSqlError>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
    , model(new QStringListModel(this))
    , attemptCount(0)  // –°—á—ë—Ç—á–∏–∫ –ø–æ–ø—ã—Ç–æ–∫
{
    ui->setupUi(this);
    ui->listViewHistory->setModel(model);
    // –ö–Ω–æ–ø–∫–∞ —Å–º–µ–Ω—ã —Ç–µ–º—ã
    connect(ui->pushButtonToggleTheme, &QPushButton::clicked, this, [=]() {
        setDarkTheme(!isDarkTheme);
        ui->pushButtonStartGame->setStyleSheet("");
        ui->pushButtonStartGame->setFixedSize(500, 40);
    });

    // –ö–Ω–æ–ø–∫–∞ –ø–µ—Ä–µ–∫–ª—é—á–µ–Ω–∏—è —è–∑—ã–∫–∞ (—Ç–æ–ª—å–∫–æ –≤ –≥–ª–∞–≤–Ω–æ–º –º–µ–Ω—é)
    connect(ui->pushButtonToggleLanguage, &QPushButton::clicked, this, [=]() {
        isEnglish = !isEnglish;
        switchLanguage(isEnglish);
    });

    setDarkTheme(true);

    // –°–∫—Ä—ã–≤–∞–µ–º –∏–≥—Ä–æ–≤—ã–µ —ç–ª–µ–º–µ–Ω—Ç—ã –¥–æ —Å—Ç–∞—Ä—Ç–∞
    ui->pushButtonGuess->setVisible(false);
    ui->pushButtonReset->setVisible(false);
    ui->lineEditInput->setVisible(false);
    ui->labelInstruction->setVisible(false);
    ui->labelFeedback->setVisible(false);
    ui->labelDivisibility->setVisible(false);
    ui->listViewHistory->setVisible(false);
    // –ü–æ–¥–∫–ª—é—á–∞–µ–º –±–∞–∑—É –¥–∞–Ω–Ω—ã—Ö
    initDatabase();
}

MainWindow::~MainWindow()
{
    if (db.isOpen())
        db.close();
    delete ui;
}

// üìå –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è –±–∞–∑—ã –¥–∞–Ω–Ω—ã—Ö SQLite
void MainWindow::initDatabase()
{
    db = QSqlDatabase::addDatabase("QSQLITE");
    db.setDatabaseName("records.db");

    if (!db.open()) {
        qDebug() << "–û—à–∏–±–∫–∞ –æ—Ç–∫—Ä—ã—Ç–∏—è –±–∞–∑—ã –¥–∞–Ω–Ω—ã—Ö:" << db.lastError().text();
        return;
    }

    // –°–æ–∑–¥–∞—ë–º —Ç–∞–±–ª–∏—Ü—É `records`, –µ—Å–ª–∏ –æ–Ω–∞ –µ—â—ë –Ω–µ —Å—É—â–µ—Å—Ç–≤—É–µ—Ç
    QSqlQuery query;
    query.exec("CREATE TABLE IF NOT EXISTS records ("
               "id INTEGER PRIMARY KEY AUTOINCREMENT, "
               "playerName TEXT, "
               "attempts INTEGER)");
}

// üìå –ó–∞–ø—É—Å–∫ –Ω–æ–≤–æ–π –∏–≥—Ä—ã
void MainWindow::on_pushButtonStartGame_clicked()
{
    bool okMin, okMax;
    int min = ui->lineEditMin->text().toInt(&okMin);
    int max = ui->lineEditMax->text().toInt(&okMax);

    if (!okMin || !okMax || min >= max) {
        QMessageBox::warning(this, tr("–ù–µ–∫–æ—Ä—Ä–µ–∫—Ç–Ω—ã–π –¥–∏–∞–ø–∞–∑–æ–Ω"),
                             tr("–í–≤–µ–¥–∏—Ç–µ –¥–≤–∞ —Ä–∞–∑–Ω—ã—Ö —á–∏—Å–ª–∞, –≥–¥–µ –ø–µ—Ä–≤–æ–µ –º–µ–Ω—å—à–µ –≤—Ç–æ—Ä–æ–≥–æ."));
        return;
    }

    minValue = min;
    maxValue = max;
    attemptCount = 0;  // –°–±—Ä–∞—Å—ã–≤–∞–µ–º –∫–æ–ª–∏—á–µ—Å—Ç–≤–æ –ø–æ–ø—ã—Ç–æ–∫

    // –°–∫—Ä—ã–≤–∞–µ–º –Ω–µ–Ω—É–∂–Ω—ã–µ —ç–ª–µ–º–µ–Ω—Ç—ã
    ui->lineEditMin->setVisible(false);
    ui->lineEditMax->setVisible(false);
    ui->labelMin->setVisible(false);
    ui->labelMax->setVisible(false);
    ui->pushButtonStartGame->setVisible(false);
    ui->pushButtonToggleTheme->setVisible(false);
    ui->pushButtonToggleLanguage->setVisible(false); // –°–∫—Ä—ã–≤–∞–µ–º —è–∑—ã–∫ –≤ –∏–≥—Ä–æ–≤–æ–º —Ä–µ–∂–∏–º–µ
    ui->pushButtonShowRecords->setVisible(false);
    ui->pushButtonResetRecords->setVisible(false);
    // –ü–æ–∫–∞–∑—ã–≤–∞–µ–º –∏–≥—Ä–æ–≤—ã–µ —ç–ª–µ–º–µ–Ω—Ç—ã
    ui->pushButtonGuess->setVisible(true);
    ui->pushButtonReset->setVisible(true);
    ui->lineEditInput->setVisible(true);
    ui->labelInstruction->setVisible(true);
    ui->labelFeedback->setVisible(true);
    ui->labelDivisibility->setVisible(true);
    ui->listViewHistory->setVisible(true);

    startNewGame();
}

void MainWindow::startNewGame()
{
    targetNumber = QRandomGenerator::global()->bounded(minValue, maxValue + 1);
    guessHistory.clear();
    updateHistory();
    ui->labelFeedback->setText(tr("–í–≤–µ–¥–∏—Ç–µ –≤–∞—à—É –ø–µ—Ä–≤—É—é –¥–æ–≥–∞–¥–∫—É"));
    ui->labelDivisibility->clear();
    ui->lineEditInput->clear();
    ui->lineEditInput->setFocus();
}

void MainWindow::updateHistory()
{
    model->setStringList(guessHistory);
}

void MainWindow::on_pushButtonGuess_clicked()
{
    bool ok;
    int guess = ui->lineEditInput->text().toInt(&ok);
    attemptCount++;  // –£–≤–µ–ª–∏—á–∏–≤–∞–µ–º —Å—á—ë—Ç—á–∏–∫ –ø–æ–ø—ã—Ç–æ–∫

    if (!ok || guess < minValue || guess > maxValue) {
        QMessageBox::warning(this, tr("–û—à–∏–±–∫–∞ –≤–≤–æ–¥–∞"),
                             tr("–í–≤–µ–¥–∏—Ç–µ —Ü–µ–ª–æ–µ —á–∏—Å–ª–æ –æ—Ç %1 –¥–æ %2")
                                 .arg(minValue).arg(maxValue));
        return;
    }

    guessHistory.append(tr("–ü–æ–ø—ã—Ç–∫–∞: %1").arg(guess));
    updateHistory();

    if (guess < targetNumber) {
        ui->labelFeedback->setText(tr("–ó–∞–≥–∞–¥–∞–Ω–Ω–æ–µ —á–∏—Å–ª–æ –±–æ–ª—å—à–µ"));
    } else if (guess > targetNumber) {
        ui->labelFeedback->setText(tr("–ó–∞–≥–∞–¥–∞–Ω–Ω–æ–µ —á–∏—Å–ª–æ –º–µ–Ω—å—à–µ"));
    } else {
        // üèÜ –ò–≥—Ä–∞ –æ–∫–æ–Ω—á–µ–Ω–∞ ‚Äì –∑–∞–ø–∏—Å—ã–≤–∞–µ–º —Ä–µ–∫–æ—Ä–¥
        insertRecord("Player", attemptCount);
        ui->labelFeedback->setText(tr("üéâ –í—ã —É–≥–∞–¥–∞–ª–∏ —á–∏—Å–ª–æ –∑–∞ %1 –ø–æ–ø—ã—Ç–æ–∫!").arg(attemptCount));
        ui->labelDivisibility->setText(tr("–ò–≥—Ä–∞ –∑–∞–≤–µ—Ä—à–µ–Ω–∞"));
    }

    ui->lineEditInput->clear();
}

void MainWindow::resetToStartMenu()
{
    // –°–∫—Ä—ã–≤–∞–µ–º –∏–≥—Ä–æ–≤—ã–µ —ç–ª–µ–º–µ–Ω—Ç—ã
    QList<QWidget*> gameWidgets = {
        ui->pushButtonGuess, ui->pushButtonReset, ui->lineEditInput,
        ui->labelInstruction, ui->labelFeedback, ui->labelDivisibility,
        ui->listViewHistory
    };

    for (auto widget : gameWidgets) {
        widget->setVisible(false);
    }

    // –ü–æ–∫–∞–∑—ã–≤–∞–µ–º —Å—Ç–∞—Ä—Ç–æ–≤—ã–µ —ç–ª–µ–º–µ–Ω—Ç—ã
    QList<QWidget*> startWidgets = {
        ui->lineEditMin, ui->lineEditMax, ui->labelMin,
        ui->labelMax, ui->pushButtonStartGame
    };

    for (auto widget : startWidgets) {
        widget->setVisible(true);
    }

    ui->pushButtonToggleTheme->setVisible(true);
    ui->pushButtonToggleLanguage->setVisible(true);
    ui->pushButtonShowRecords->setVisible(true);
    ui->lineEditMin->clear();
    ui->lineEditMax->clear();
}

// üìå –î–æ–±–∞–≤–ª–µ–Ω–∏–µ –∑–∞–ø–∏—Å–∏ –≤ –ë–î
void MainWindow::insertRecord(const QString &playerName, int attempts)
{
    QSqlQuery query;
    query.prepare("INSERT INTO records (playerName, attempts) VALUES (?, ?)");
    query.addBindValue(playerName);
    query.addBindValue(attempts);

    if (!query.exec()) {
        qDebug() << "–û—à–∏–±–∫–∞ –∑–∞–ø–∏—Å–∏ –≤ –ë–î:" << query.lastError().text();
    }
}

// üìå –û—Ç–æ–±—Ä–∞–∂–µ–Ω–∏–µ –ª—É—á—à–∏—Ö —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–≤
void MainWindow::showRecords()
{
    QSqlQuery query("SELECT playerName, attempts FROM records ORDER BY attempts ASC LIMIT 10");
    QString recordText = tr("–¢–æ–ø —Ä–µ–∫–æ—Ä–¥–æ–≤:\n");

    while (query.next()) {
        recordText += query.value(0).toString() + ": " + QString::number(query.value(1).toInt()) + "\n";
    }

    QMessageBox::information(this, tr("–†–µ–∫–æ—Ä–¥—ã"), recordText);
}
void MainWindow::setDarkTheme(bool enabled)
{
    isDarkTheme = enabled;
    QPalette palette;
    if (enabled) {
        // –¢—ë–º–Ω–∞—è —Ç–µ–º–∞
        palette.setColor(QPalette::Window, QColor(35, 35, 35));
        palette.setColor(QPalette::WindowText, QColor(220, 220, 220));
        palette.setColor(QPalette::Base, QColor(42, 42, 42));
        palette.setColor(QPalette::AlternateBase, QColor(66, 66, 66));
        palette.setColor(QPalette::Button, QColor(53, 53, 53));
        palette.setColor(QPalette::ButtonText, Qt::white);
        palette.setColor(QPalette::Highlight, QColor(42, 130, 218));
        palette.setColor(QPalette::HighlightedText, Qt::black);
    } else {
        // –°–≤–µ—Ç–ª–∞—è —Ç–µ–º–∞ (—á–∏—Å—Ç–æ –±–µ–ª—ã–π —Ñ–æ–Ω)
        palette.setColor(QPalette::Window, QColor("#FFFFFF"));  // –ë–µ–ª—ã–π —Ñ–æ–Ω
        palette.setColor(QPalette::WindowText, QColor("#000000"));  // –ß—ë—Ä–Ω—ã–π —Ç–µ–∫—Å—Ç
        palette.setColor(QPalette::Base, QColor("#FFFFFF"));  // –ü–æ–ª—è –≤–≤–æ–¥–∞ - –±–µ–ª—ã–µ
        palette.setColor(QPalette::AlternateBase, QColor("#F0F0F0"));
        palette.setColor(QPalette::Button, QColor("#E6E6E6"));
        palette.setColor(QPalette::ButtonText, QColor("#000000"));
        palette.setColor(QPalette::Highlight, QColor("#007ACC"));
        palette.setColor(QPalette::HighlightedText, QColor("#FFFFFF"));
    }
    qApp->setPalette(palette);
}

void MainWindow::switchLanguage(bool toEnglish)
{
    qApp->removeTranslator(&translator);
    if (toEnglish) {
        QString translationFile = QCoreApplication::applicationDirPath() + "/translations/app_en.qm";
        if (translator.load(translationFile))
            qApp->installTranslator(&translator);
    }
    ui->retranslateUi(this);
    this->update();
    qApp->processEvents();
}
void MainWindow::on_pushButtonReset_clicked()
{
    resetToStartMenu();
}
void MainWindow::on_pushButtonShowRecords_clicked()
{
    showRecords();
}
void MainWindow::on_pushButtonResetRecords_clicked()
{
    QMessageBox::StandardButton reply;
    reply = QMessageBox::question(this, tr("–°–±—Ä–æ—Å —Ä–µ–∫–æ—Ä–¥–æ–≤"),
                                  tr("–í—ã —É–≤–µ—Ä–µ–Ω—ã, —á—Ç–æ —Ö–æ—Ç–∏—Ç–µ —É–¥–∞–ª–∏—Ç—å –≤—Å–µ —Ä–µ–∫–æ—Ä–¥—ã?"),
                                  QMessageBox::Yes | QMessageBox::No);

    if (reply == QMessageBox::Yes) {
        QSqlQuery query;
        if (!query.exec("DELETE FROM records")) {
            qDebug() << "–û—à–∏–±–∫–∞ –æ—á–∏—Å—Ç–∫–∏ –±–∞–∑—ã –¥–∞–Ω–Ω—ã—Ö:" << query.lastError().text();
        } else {
            QMessageBox::information(this, tr("–°–±—Ä–æ—Å —Ä–µ–∫–æ—Ä–¥–æ–≤"), tr("–í—Å–µ —Ä–µ–∫–æ—Ä–¥—ã —É–¥–∞–ª–µ–Ω—ã."));
        }
    }
}
